src/bitcoin-chainstate.cpp

```diff
@@ -80,12 +82,22 @@ int main(int argc, char* argv[])
 
     GetMainSignals().RegisterBackgroundSignalScheduler(scheduler);
 
+    class KernelNotifications : public kernel::Notifications
+    {
+    public:
+        void blockTip(SynchronizationState, CBlockIndex&) override
```

Comment:
Would it be possible to keep pointers/references to internal data out of this interface from the beginning? I'm not sure if that's a goal at this layer.

For blocktip for example, taking a quick look around, for the ui it looks like this eventually gets converted to:
`void(SynchronizationState, interfaces::BlockTip tip, double verification_progress)`

I wonder if we can do the same here?

Reply:
Primarily, I did not want to mess with existing interfaces, but I checked through everything and it would be feasibly to entirely drop usage of `CBlockIndex` for this notification. The change is a bit sprawling though and also ends up touching rpc code.

>  I'm not sure if that's a goal at this layer.

The way I see this play out during the next phase of exploring a more suitable kernel api, these notifications might be used to feed back into the `ChainstateManager`. As long as the chainman interface primarily takes a `CBlockIndex` such a change feels like putting the cart before the horse to me.

Reply:
Sure. I only asked because it looked like it was possible in this case. But this can be done after-the-fact if we decide that's what we want.

Longer-term, we'd want to detach from our internal structures to avoid changing the abi every time we change something internally. So I don't imagine `CBlockIndex` ever being part of a public api. Maybe a trivial wrapper or something, but not CBlockIndex itself. (It's also worth keeping that in mind at this stage because that external-to-internal-component translation becomes part of the overhead of using the api)

Reply:
Will add this to the TODO section in the tracking issue.